### PR TITLE
feat(session replay): delay compresion 

### PR DESCRIPTION
**Pull request by jxiwang**
### Summary

Delays the compression of events emitted by rrrweb record using requestIdleCallback.

### Checklist

* [ ] Does your PR title have the correct [title format](https://github.com/amplitude/Amplitude-TypeScript/blob/main/CONTRIBUTING.md#pr-commit-title-conventions)?
* Does your PR have a breaking change?:

### CODE REVIEW

**Review by lewgordon-amplitude — CHANGES_REQUESTED** [review]
This is really cool. I had no idea about `requestIdleCallback` before the PR, it seems like a good use case for this. I'm wondering though if we want to change this a little bit. The things that come to mind for me:

1. The hard coded timeout seems like it'll fire a lot of events just slightly delayed. Therefore not immediately impacting performance, but just having it delayed. I think it being hard coded is to going to be slightly problematic since that number will vary from app to app.
2. It seems like we're creating a new callback for each event. I'm not entirely certain it's a problem, but for a high number of events this could be a lot of callbacks.

I'm thinking to help mitigate some of these things maybe we could instead of a queue that we work through during idle periods and get through as many as possible. We could probably use the [IdleDeadline](https://developer.mozilla.org/en-US/docs/Web/API/IdleDeadline) object to know when to quit. The idea being that we register a single callback that processes through the queue as much as possible during idle periods. Getting this exactly right and not going over time may be tough, but I think this would be more performant.

I think in either case before we merge this we should add an option so that we can test this ourselves for now and make sure we add some performance charts with a sample app or preferably our site so we're confident in the change.

I'm definitely open to feedback/other thoughts on this!

File: packages/session-replay-browser/src/session-replay.ts [hunk 1]
```diff
@@ -309,6 +310,36 @@ export class SessionReplay implements AmplitudeSessionReplay {
...
+    const deviceId = this.getDeviceId();
+    this.eventsManager &&
+      deviceId &&
+      this.eventsManager.addEvent({ event: { type: 'replay', data: compressedEvent }, sessionId, deviceId });
+  };
+
+  deferEventCompression = (event: eventWithTime, sessionId: number) => {
+    const globalScope = getGlobalScope();
+    // In case the browser does not support requestIdleCallback, we will compress the event immediately
+    if (globalScope && 'requestIdleCallback' in globalScope) {
```
[the remark this answers is not in the document]
Comment (jxiwang) [reply packages/session-replay-browser/src/session-replay.ts:330]: Ah true, the browser definitely wouldn't change. Let me see if I can move the check out of this function, maybe just check prior to recording.

**Comment by jxiwang** [comment]
@lewgordon-amplitude 

Thanks for the feedback! 

1. For the timeout, I considered not including one but am also concerned with the case that the compression doesn't get fired. Even if the timeout is reached, the compression _should_ get pushed to the back of the event queue in cases like when the page takes > 2s to load. Although I'm not entirely sure that doesn't have performance implications still. I could remove the timeout, and we could monitor the effects. 

2. Good call on the callbacks being generated per event. I'll add an event queue based on the `timeRemaining`, and I actually did not know `idleDeadline` could provide that information. Pretty cool that it does do that! 

Agreed on adding an option so we can have control over testing this, and we can let customers know this is something they could turn on.

File: packages/session-replay-browser/src/session-replay.ts [hunk 2]
```diff
@@ -312,12 +315,12 @@ export class SessionReplay implements AmplitudeSessionReplay {
   recordEvents() {
     const shouldRecord = this.getShouldRecord();
     const sessionId = this.identifiers?.sessionId;
+    // const canDelayCompression = globalScope && 'requestIdleCallback' in globalScope;
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/session-replay.ts:318]:
```suggestion
```

No longer needed?

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 3]
```diff
@@ -0,0 +1,95 @@
...
+
+  // Schedule processing during idle time
+  public scheduleIdleProcessing(): void {
+    if (!this.isProcessing) {
+      this.isProcessing = true;
+      requestIdleCallback(
+        (idleDeadline) => {
+          this.processQueue(idleDeadline);
+        },
+        { timeout: 2000 },
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/event-compressor.ts:40]: I think this should be a parameter in the config, too.

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 4]
```diff
@@ -0,0 +1,95 @@
...
+      );
+    } else {
+      this.isProcessing = false;
+    }
+  }
+
+  compressEvents = (event: eventWithTime) => {
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/event-compressor.ts:79]:
```suggestion
  compressEvent = (event: eventWithTime) => {
```

Nit: :)

File: packages/session-replay-browser/src/events/event-compressor.ts [hunk 5]
```diff
@@ -0,0 +1,95 @@
...
+  sessionId: number;
+}
+
+export class EventCompressor {
+  taskQueue: TaskQueue[] = [];
+  isProcessing = false;
+  eventsManager?: SessionReplayEventsManager<'replay' | 'interaction', string>;
+  config: SessionReplayJoinedConfig;
+  deviceId: string | undefined;
+  canUseIdleCallback: boolean | undefined;
```
Comment (lewgordon-amplitude) [inline packages/session-replay-browser/src/events/event-compressor.ts:18]: I think the code could be cleaner if these instance variables were not possibly undefined. It seems like all are set in the constructor, so I don't think we need to recheck them in the class methods.

**Review by lewgordon-amplitude — APPROVED** [review]
LGTM! I think there are minor changes we can make, but with the config in place and the current approach I don't see anything majorly blocking here.

**Review by lewgordon-amplitude — APPROVED** [review]

**Pull request merged**